SzenenSteuerung/module.php

```diff
@@ -1,75 +1,107 @@
+	public function Create()
+	{
 		//Never delete this line!
 		parent::Create();
 
 		//Properties
-		$this->RegisterPropertyInteger("SceneCount", 3);
-		
-		if(!IPS_VariableProfileExists("SZS.SceneControl")){
+		$this->RegisterPropertyInteger("SceneCount", 5);
+		$this->RegisterAttributeString("SceneData", "[]");
+
+
+		if (!IPS_VariableProfileExists("SZS.SceneControl")) {
 			IPS_CreateVariableProfile("SZS.SceneControl", 1);
 			IPS_SetVariableProfileValues("SZS.SceneControl", 1, 2, 0);
 			//IPS_SetVariableProfileIcon("SZS.SceneControl", "");
 			IPS_SetVariableProfileAssociation("SZS.SceneControl", 1, "Speichern", "", -1);
 			IPS_SetVariableProfileAssociation("SZS.SceneControl", 2, "Ausführen", "", -1);
 		}
-
 	}
 
-	public function Destroy() {
+	public function Destroy()
+	{
 		//Never delete this line!
 		parent::Destroy();
 	}
 
-	public function ApplyChanges() {
+	public function ApplyChanges()
+	{
 		//Never delete this line!
 		parent::ApplyChanges();
-		
+
 		$this->CreateCategoryByIdent($this->InstanceID, "Targets", "Targets");
-		
-		for($i = 1; $i <= $this->ReadPropertyInteger("SceneCount"); $i++) {
-			if(@IPS_GetObjectIDByIdent("Scene".$i, $this->InstanceID) === false){
-				//Scene
-				$vid = IPS_CreateVariable(1 /* Scene */);
-				IPS_SetParent($vid, $this->InstanceID);
-				IPS_SetName($vid, "Scene".$i);
-				IPS_SetIdent($vid, "Scene".$i);
-				IPS_SetVariableCustomProfile($vid, "SZS.SceneControl");
-				$this->EnableAction("Scene".$i);
-				SetValue($vid, 2);
-				//SceneData
-				$vid = IPS_CreateVariable(3 /* SceneData */);
-				IPS_SetParent($vid, $this->InstanceID);
-				IPS_SetName($vid, "Scene".$i."Data");
-				IPS_SetIdent($vid, "Scene".$i."Data");
-				IPS_SetHidden($vid, true);
-				
+
+		for ($i = 1; $i <= $this->ReadPropertyInteger("SceneCount"); $i++) {
+			$variableID = $this->RegisterVariableInteger("Scene" . $i, "Scene" . $i, "SZS.SceneControl");
+			$this->EnableAction("Scene" . $i);
+			SetValue($variableID, 2);
+		}
+
+		for ($k = 1; $k <= $this->ReadPropertyInteger("SceneCount"); $k++) {
+			$SceneDataID = @$this->GetIDForIdent("Scene" . $i . "Data");
+			if ($SceneDataID && function_exists("wddx_deserialize")) {
+
+				$data = wddx_deserialize(GetValue($SceneDataID));
+				if ($data !== NULL) {
+					SetValue($SceneDataID, json_encode($data));
+				}
 			}
 		}
+		$SceneData = json_decode($this->ReadAttributeString("SceneData"));
 
-		//Delete excessive Scences 
-		$ChildrenIDsCount = sizeof(IPS_GetChildrenIDs($this->InstanceID))/2;
-		if($ChildrenIDsCount > $this->ReadPropertyInteger("SceneCount")) {
-			for($j = $this->ReadPropertyInteger("SceneCount")+1; $j <= $ChildrenIDsCount; $j++) {
-				IPS_DeleteVariable(IPS_GetObjectIDByIdent("Scene".$j, $this->InstanceID));
-				IPS_DeleteVariable(IPS_GetObjectIDByIdent("Scene".$j."Data", $this->InstanceID));
+		if (!is_array($SceneData)) {
+			$SceneData = [];
+		}
+
+		for ($i = 1; $i <= $this->ReadPropertyInteger("SceneCount"); $i++) {
```

Comment:
Ich würde bevorzugen, dass wir die Logik zum "Importieren" und zum "Initialisieren" trennen. Das hat den Vorteil, dass du die zwei Dinge klar voneinander trennst. Und es macht es auf lange Sicht einfacher zu verstehen was passiert.

Ein paar Kommentare wären nett. z.B. 

//Import from WDDX data into new JSON data
//Import JSON Data from Data Variables into Attribute and cleanup Data Variables
//Initialize SceneData